### ISSUE TITLE
Q: Regarding vanishing leader in Postgres

### ISSUE THREAD

**Issue by agis** [drio-ai/leaderelection#2]
In the README it's stated that:

> Postgres implementation uses session advisory locking to run the election. In case a leader vanishes, Postgres will have to detect this scenario before the lock is released as a result of cleaning up the associated session. The sooner Postgres detects a leader is gone, the sooner it will cleanup and a new leader will be elected. **There are multiple ways to do this and it is left to the user to figure out a strategy that best fits their environment.**

Can you elaborate on the highlighted sentence from the quote? I was thinking that setting `tcp_keepalives_*` and `client_connection_check_interval` settings in Postgres will alleviate such issues, since the server would detect quickly that a connection is hung and terminate it.

Thanks in advance!

**Comment by sraghavan-drio** [comment]
My experience with Postgres is limited. The TCP keepalive method is what I used in my setup to fix this problem. I was just not sure if this is the recommended approach and, if there are other ways that seasoned Postgres users know to address this. This is the reason for the statement in bold.

**Comment by agis** [comment]
In the current implementation, is there a heartbeat mechanism which ensures that the lease will be relinquished even if there's a network partition between the leader and the database?

**Comment by sraghavan-drio** [comment]
> In the current implementation, is there a heartbeat mechanism which ensures that the lease will be relinquished even if there's a network partition between the leader and the database?

Yes, there is. For a leader it is LeaderCheckInterval and follower it is FollowerCheckInterval. Please note, distributed leader election works on the principle of a trusted third party being the arbiter. In the case of this module, this can either be Postgres advisory lock or Redis distributed lock. In case of a network issue, it is important for the third party to detect this scenario in order to elect a new leader. This is done using Postgres keepalive timeout or Redis lock TTL.

As far as what should happen on the service using this module, it is up to implementation. The current module implementation will result in an error and the election goroutine exiting. This could be trigger for the caller to exit and respawn. Like we have discussed in another issue, this might not always be the right choice. I will factor in a failure count or something similar to handle transient network issues.

**Comment by agis** [comment]
> The TCP keepalive method is what I used in my setup to fix this problem.

Can you elaborate on how exactly you achieved this? Which Postgres settings did you use?

**Comment by sraghavan-drio** [comment]
> > The TCP keepalive method is what I used in my setup to fix this problem.
> 
> Can you elaborate on how exactly you achieved this? Which Postgres settings did you use?

This is how I run postgres docker container

```
 configdb:
  image: postgres:17.2
  container_name: postgres
  command:
    - -ctcp_keepalives_idle=20
    - -ctcp_keepalives_interval=3
    - -ctcp_keepalives_count=3
```

**Comment by sraghavan-drio** [comment]
Closing issue. Please reopen if you have any further questions.
